cmd/util/ledger/migrations/storage_used_update_migration.go

```diff
@@ -0,0 +1,216 @@
+package migrations
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	fvm "github.com/onflow/flow-go/fvm/state"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/utils"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+type accountPayloadSize struct {
+	Address     string
+	StorageUsed uint64
+}
+
+type indexedPayload struct {
+	Index   int
+	Payload ledger.Payload
+}
+
+type accountStorageUsedPayload struct {
+	Address string
+	Index   int
+}
+
+type StorageUsedUpdateMigration struct {
+	Log       zerolog.Logger
+	OutputDir string
+}
+
+func (m *StorageUsedUpdateMigration) filename() string {
+	return path.Join(m.OutputDir, fmt.Sprintf("storage_used_update_%d.csv", int32(time.Now().Unix())))
+}
+
+// iterates through registers keeping a map of register sizes
+// after it has reached the end it add storage used and storage capacity for each address
+func (m *StorageUsedUpdateMigration) Migrate(payload []ledger.Payload) ([]ledger.Payload, error) {
+	fn := m.filename()
+	m.Log.Info().Msgf("Running Storage Used update. Saving output to %s.", fn)
+
+	f, err := os.Create(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err = f.Close()
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	writer := bufio.NewWriter(f)
+	defer func() {
+		err = writer.Flush()
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	storageUsed := make(map[string]uint64)
+	storageUsedChan := make(chan accountPayloadSize, 1)
+	payloadChan := make(chan indexedPayload)
+	storageUsedPayloadChan := make(chan accountStorageUsedPayload, 1)
+	storageUsedPayload := make(map[string]int)
+
+	workerCount := runtime.NumCPU()
+	inputWG := &sync.WaitGroup{}
+	outputWG := &sync.WaitGroup{}
+
+	outputWG.Add(1)
+	go func() {
+		for payloadSize := range storageUsedChan {
+			if _, ok := storageUsed[payloadSize.Address]; !ok {
+				storageUsed[payloadSize.Address] = 0
+			}
+			storageUsed[payloadSize.Address] = storageUsed[payloadSize.Address] + payloadSize.StorageUsed
```

Comment:
does this assumes that an address exist that has two storage_used registers?

Reply:
No.

the storageUsedChan gets 1 input for every single register. That input is the address of a register and the storage used by that register. That is then just summed up to get the storage used per address.

This is basically just a way to prevent concurrent access to the `storageUsed` dictionary.